keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,111 @@
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-05-14
+last-updated: 2018-05-14
+status: provisional
+---
+
+# Sidecar Containers
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives-optional)
+
+## Summary
+
+To solve the issue of sidecars not being well supported we can create a new type of container called a sidecar which are handled like normal containers but with the caveat that a normal container's status could take precedence over the sidecar's status. This will solve the problem of Jobs not completing/failing when a sidecar container never exits due to it having no awareness of the 'primary' container.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish. The only way around this is to build some kind of logic into the sidecar so that it will exit when the other container exits. This can be quite difficult if you're using a sidecar provided by a third party and results in some nasty hacks being implemented to achieve this.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a normal container in almost all respects apart from the fact that in certain situations their status will be ignored:
+* When all normal containers are in `Completed` State
+* When all normal containers are in `Error` State
+In a job for example this means that you can have as many sidecars as you like and they will all get terminated when the normal containers all Complete or if all normal containers error.
```

Comment:
- `s/normal/non-sidecar`

Another way of wording the behavior is that running sidecar containers are ignored when determining if the pod has terminated (with or without an error); when all non-sidecar containers terminate, all running sidecar containers are terminated, but flagged so that they appear as having terminated successfully. This is slightly different than above in that sidecars that terminate in failure (with a restart policy of never) cause the entire pod to transition to the `Failed` phase, i.e. a pod still fails if a sidecar fails.

Reply:
ah yeah I see what you mean, so currently if a sidecar were to fail and all the other containers completed, the pod phase would be `Succeeded` which is quite misleading as something failed. I'll reword this section, thanks.